### ISSUE TITLE
assert callable(unaltered_fn)

### ISSUE THREAD

**Issue by ezyang** [pytorch/pytorch#98434]
### 🐛 Describe the bug

This is a bug generated from https://github.com/pytorch/pytorch/issues/97078

To reproduce, check out transformers and patch (I tested on a515d0a77c769954ac2f0151a2a99c04d8d6cf95)

```
diff --git a/src/transformers/trainer.py b/src/transformers/trainer.py
index 2eb081af7..886df74c1 100755
--- a/src/transformers/trainer.py
+++ b/src/transformers/trainer.py
@@ -1572,8 +1572,7 @@ class Trainer:
 
         # torch.compile() needs to be called after wrapping the model with FSDP or DDP
         # to ensure that it accounts for the graph breaks required by those wrappers
-        if self.args.torch_compile:
-            model = torch.compile(model, backend=self.args.torch_compile_backend, mode=self.args.torch_compile_mode)
+        model = torch.compile(model, backend=self.args.torch_compile_backend, mode=self.args.torch_compile_mode, dynamic=True)
 
         return model
 
```

Then, run

```
pytest tests/trainer/test_trainer.py --tb=native -k test_adafactor_lr_none
```

It fails with

```
______________________________________________ TrainerIntegrationPrerunTest.test_adafactor_lr_none ______________________________________________
Traceback (most recent call last):
  File "/data/users/ezyang/a/transformers/tests/trainer/test_trainer.py", line 465, in setUp
    trainer.train()
  File "/data/users/ezyang/a/transformers/src/transformers/trainer.py", line 1658, in train
    return inner_training_loop(
  File "/data/users/ezyang/a/transformers/src/transformers/trainer.py", line 1745, in _inner_training_loop
    model = self._wrap_model(self.model_wrapped)
  File "/data/users/ezyang/a/transformers/src/transformers/trainer.py", line 1575, in _wrap_model
    model = torch.compile(model, backend=self.args.torch_compile_backend, mode=self.args.torch_compile_mode, dynamic=True)
  File "/data/users/ezyang/a/pytorch/torch/__init__.py", line 1600, in compile
    return torch._dynamo.optimize(backend=backend, nopython=fullgraph, dynamic=dynamic, disable=disable)(model)
  File "/data/users/ezyang/a/pytorch/torch/_dynamo/eval_frame.py", line 499, in optimize
    return _optimize_catch_errors(
  File "/data/users/ezyang/a/pytorch/torch/_dynamo/eval_frame.py", line 401, in _optimize_catch_errors
    return OptimizeContext(
  File "/data/users/ezyang/a/pytorch/torch/_dynamo/eval_frame.py", line 330, in __init__
    compiler_fn = innermost_fn(callback)
  File "/data/users/ezyang/a/pytorch/torch/_dynamo/eval_frame.py", line 149, in innermost_fn
    assert callable(unaltered_fn)
AssertionError
```

It's possible that HF is misusing the torch.compile API (there's some sort of repeated wrapping going on, it seems like), but even if that's true, it shouldn't assert error.

cc @gchanan @zou3519 @soumith @msaroufim @wconstab @ngimel @bdhirsh @stas00 

### Versions

master

**Comment by stas00** [comment]
@ezyang, the model wrapping is just to deal with with DDP or Deepspeed or whatever other framework drives the multi-gpu training. I'm not sure where you see a repeated wrapping.

The wrapping code is here:
https://github.com/huggingface/transformers/blob/15641892985b1d77acc74c9065c332cd7c3f7d7f/src/transformers/trainer.py#L1386

**Comment by voznesenskym** [comment]
@mlazos can you confirm?

**Comment by mlazos** [comment]
I'm unable to repro this - it doesn't look like HF calls torch.compile in their trainer wrapper code anymore. I grepped the repo too and couldn't find where the torch_compile trainer args are used.

**Comment by penguinwu** [comment]
Close since Lazos cannot repro. @ezyang Feel free to reopen w/ repro instruction if still fails for you.

**Comment by stas00** [comment]
@mlazos, I investigated this - `torch.compile` functionality was delegated to Accelerate in this PR:
https://github.com/huggingface/transformers/commit/03db59104714f60b591ef48073840b288ee4cdc0
so it's still there in the trainer, but is now done by Accelerate if `--torch_compile` flag is passed.

**Comment by stas00** [comment]
I was able to update everything and I confirm that this particular issue is no longer there.

I still can't use `torch.compile` - even after turning off DDP `torch._dynamo.config.optimize_ddp=False`

it spins the CPU - looking at the stack trace it's deep inside recursive `percolate_tags` calls - fluctuating between 100-150 frames deep, I gave up after 10min. This was just llama2 model from HF transformers.

Well, will try again in 6 months.
